### ISSUE TITLE
[JENKINS-36439] Evolve more composable redux store code patterns

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#2966]
<p>This is assigned to <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a> for now, but it's really <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tscherler" class="user-hover" rel="tscherler">tscherler</a> + <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a>, working together to evolve the current redux codebase to something that is easier to maintain, test, extend etc.</p>

<p>We are using redux more and more and it works well (it's fundamental to how BlueO works), but we have stretched the current implementation beyond the point where it is easy/possible for others to make sense of it, contribute to it etc. It's getting harder and harder to maintain, so we really need to evolve it to the next stage.</p>

<p>There may be other evolutionary steps required/possible (e.g. generating from the backend rest api) as the whole app evolves.</p>

<p>We'll see if we can come up with a more composable pattern for managing app state in redux. Maybe something like a <tt>StoreObject</tt> JS type that can be added as an extension point (in the <tt>.yaml</tt> file) that encapsulates a redux reducer for handling a specific set of redux actions relating to a specific piece of state (e.g. branches on a job) + providing helpers for loading/fetching data from the REST API etc.</p>

<p>Let's see about <tt>Immutablejs</tt> too. <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tscherler" class="user-hover" rel="tscherler">tscherler</a> feels strongly that this is the right way to store the state. He has very valid points re reference Vs value equality etc and how Immutablejs helps there + management of large amounts of data (freezing and cloning the whole state will not scale as we grow), but others find Immutablejs a bit cumbersome to use (myself included). Maybe there's a happy medium..</p>

<p>Moving everything to a new pattern in one go would be a big task, so it would be good if the new pattern could live alongside the existing stuff as we transition. That way, we can divide and conquer.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly">tfennelly</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439" target="_blank">Evolve more composable redux store code patterns</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly">tfennelly</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: technical-debt
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2016-07-27T07:52:56+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 3
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>This is assigned to <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a> for now, but it's really <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tscherler" class="user-hover" rel="tscherler">tscherler</a> + <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a>, working together to evolve the current redux codebase to something that is easier to maintain, test, extend etc.</p>

<p>We are using redux more and more and it works well (it's fundamental to how BlueO works), but we have stretched the current implementation beyond the point where it is easy/possible for others to make sense of it, contribute to it etc. It's getting harder and harder to maintain, so we really need to evolve it to the next stage.</p>

<p>There may be other evolutionary steps required/possible (e.g. generating from the backend rest api) as the whole app evolves.</p>

<p>We'll see if we can come up with a more composable pattern for managing app state in redux. Maybe something like a <tt>StoreObject</tt> JS type that can be added as an extension point (in the <tt>.yaml</tt> file) that encapsulates a redux reducer for handling a specific set of redux actions relating to a specific piece of state (e.g. branches on a job) + providing helpers for loading/fetching data from the REST API etc.</p>

<p>Let's see about <tt>Immutablejs</tt> too. <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tscherler" class="user-hover" rel="tscherler">tscherler</a> feels strongly that this is the right way to store the state. He has very valid points re reference Vs value equality etc and how Immutablejs helps there + management of large amounts of data (freezing and cloning the whole state will not scale as we grow), but others find Immutablejs a bit cumbersome to use (myself included). Maybe there's a happy medium..</p>

<p>Moving everything to a new pattern in one go would be a big task, so it would be good if the new pattern could live alongside the existing stuff as we transition. That way, we can divide and conquer.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=263243&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263243">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>After talking to <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> and he was asking if anything else was depending on this.</p>

<p>So, more-or-less all features are depending on redux in some way. We could keep going with redux as we have it today, so from that point of view one could say that this is not a "blocker". Otoh, we know that how we're doing redux now needs to evolve/change and is one of the areas where we are building up most tech-debt, so from my pov, I'd prefer for us to look at this sooner rather than later and see can we start the process of improving it Vs continuing to leave the "sh1t pile up in the back room" <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Of course in the end, others might consider that it's better to keep pushing forward with more features and show more visual progress.</p>
</pre>
</details>

</ul>
</details>

<p>After talking to <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> and he was asking if anything else was depending on this.</p>

<p>So, more-or-less all features are depending on redux in some way. We could keep going with redux as we have it today, so from that point of view one could say that this is not a "blocker". Otoh, we know that how we're doing redux now needs to evolve/change and is one of the areas where we are building up most tech-debt, so from my pov, I'd prefer for us to look at this sooner rather than later and see can we start the process of improving it Vs continuing to leave the "sh1t pile up in the back room" <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Of course in the end, others might consider that it's better to keep pushing forward with more features and show more visual progress.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=263328&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263328">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I think this is only a blocker in the sense that it is "very important" for immediate things that are desperately needed like paging, it isn't blocking most other features that are currently being worked on.</p>
</pre>
</details>

</ul>
</details>

<p>I think this is only a blocker in the sense that it is "very important" for immediate things that are desperately needed like paging, it isn't blocking most other features that are currently being worked on.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=263354&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263354">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Right, any feature that is not storing state in the redux store doesn't care about this.</p>

<p>Any feature that does store state is not hard blocked by this i.e. we could just keep going as-is, but imo we should not put off trying to clean up the redux code any longer as it's getting more and more unmaintainable with every new feature contributing to it.</p>
</pre>
</details>

</ul>
</details>

<p>Right, any feature that is not storing state in the redux store doesn't care about this.</p>

<p>Any feature that does store state is not hard blocked by this i.e. we could just keep going as-is, but imo we should not put off trying to clean up the redux code any longer as it's getting more and more unmaintainable with every new feature contributing to it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=263355&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263355">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a> understood, and it makes sense to make pagination depend on this stuff? (as in, not happen at the same time)? </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tfennelly" class="user-hover" rel="tfennelly">tfennelly</a> understood, and it makes sense to make pagination depend on this stuff? (as in, not happen at the same time)? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=263356&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263356">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> as I said, we could forge ahead with pagination and all other features without doing this, but that's a bad idea imo as it's just putting off the inevitable and making it a bigger job when we do it (more code to refactor).</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale" class="user-hover" rel="michaelneale">michaelneale</a> as I said, we could forge ahead with pagination and all other features without doing this, but that's a bad idea imo as it's just putting off the inevitable and making it a bigger job when we do it (more code to refactor).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=263372&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263372">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I know <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tscherler" class="user-hover" rel="tscherler">tscherler</a> mentioned perhaps ditching <tt>connect</tt> and just passing props down the entire view hierarchy - although I'm not sure if there's an automated way to do this, or if it has to be manual at each level. Connect allows us to avoid writing container components and a lot of boilerplate, but it's so terse that it's extremely hard to follow and debug. Maybe I just need more practice with it.</p>

<p>I'm setting up a new store for Personalization and trying to tweak the organization of things a little bit, so we should keep in touch about where you guys settle. The ticket FYI is <a href="https://issues.jenkins.io/browse/JENKINS-35837" title="Developer can view the dashboard" class="issue-link" data-issue-key="JENKINS-35837"><del>JENKINS-35837</del></a>.</p>

<p>I would also like to add that we should put a little thought into what we might be able to code generate later - particularly ImmutableJS classes that match against the objects sent over the wire in JSON. Don't need to write the code to do that yet, but maybe we want to structure things in such a way that each class is in its own file, and that are all imported/exported via a single manifest file. e.g.</p>

<p>model/
&#8211; index.js (exports all classes)
&#8211; Run.js
&#8211; Pipeline.js</p>
</pre>
</details>

</ul>
</details>

<p>I know <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tscherler" class="user-hover" rel="tscherler">tscherler</a> mentioned perhaps ditching <tt>connect</tt> and just passing props down the entire view hierarchy - although I'm not sure if there's an automated way to do this, or if it has to be manual at each level. Connect allows us to avoid writing container components and a lot of boilerplate, but it's so terse that it's extremely hard to follow and debug. Maybe I just need more practice with it.</p>

<p>I'm setting up a new store for Personalization and trying to tweak the organization of things a little bit, so we should keep in touch about where you guys settle. The ticket FYI is <a href="https://issues.jenkins.io/browse/JENKINS-35837" title="Developer can view the dashboard" class="issue-link" data-issue-key="JENKINS-35837"><del>JENKINS-35837</del></a>.</p>

<p>I would also like to add that we should put a little thought into what we might be able to code generate later - particularly ImmutableJS classes that match against the objects sent over the wire in JSON. Don't need to write the code to do that yet, but maybe we want to structure things in such a way that each class is in its own file, and that are all imported/exported via a single manifest file. e.g.</p>

<p>model/<br/>
&#8211; index.js (exports all classes)<br/>
&#8211; Run.js<br/>
&#8211; Pipeline.js</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=263384&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263384">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>I know Thorsten Scherler mentioned perhaps ditching connect and just passing props down the entire view hierarchy - although I'm not sure if there's an automated way to do this, or if it has to be manual at each level. Connect allows us to avoid writing container components and a lot of boilerplate, but it's so terse that it's extremely hard to follow and debug. Maybe I just need more practice with it.</p></blockquote>

<p>Yeah, I'm not at all a fan of connect. It removes so much that it's more or less impossible grok what's happening without having a PhD in redux. That's not good at all from a code maintenance pov.</p>

<p>Connect is really just a wrapper around passing the store via the react context from a top level <tt>&lt;Provider&gt;</tt>. You can do that easily enough by hand. It also handles the component store subscribe/unsubscribe to trigger the render/rerender, which is handy but can also be done easily enough by hand.</p>

<p>There are other things like this in some of these libs that I do not like at all. Some people seem to have the idea that the only thing that matters is the terseness of the code, which is a huge anti-pattern imo. I'm much more in favour of a bit more code that is easier to follow and a bit more self explanatory i.e. not requiring the PhD in redux before you can make sense of it.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>I know Thorsten Scherler mentioned perhaps ditching connect and just passing props down the entire view hierarchy - although I'm not sure if there's an automated way to do this, or if it has to be manual at each level. Connect allows us to avoid writing container components and a lot of boilerplate, but it's so terse that it's extremely hard to follow and debug. Maybe I just need more practice with it.</p></blockquote>

<p>Yeah, I'm not at all a fan of connect. It removes so much that it's more or less impossible grok what's happening without having a PhD in redux. That's not good at all from a code maintenance pov.</p>

<p>Connect is really just a wrapper around passing the store via the react context from a top level <tt><Provider></tt>. You can do that easily enough by hand. It also handles the component store subscribe/unsubscribe to trigger the render/rerender, which is handy but can also be done easily enough by hand.</p>

<p>There are other things like this in some of these libs that I do not like at all. Some people seem to have the idea that the only thing that matters is the terseness of the code, which is a huge anti-pattern imo. I'm much more in favour of a bit more code that is easier to follow and a bit more self explanatory i.e. not requiring the PhD in redux before you can make sense of it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=263388&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263388">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I agree. We need to keep the learning curve shallow so plugin authors can make sense of this stuff. We can chat more about it tomorrow.</p>
</pre>
</details>

</ul>
</details>

<p>I agree. We need to keep the learning curve shallow so plugin authors can make sense of this stuff. We can chat more about it tomorrow.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=263390&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263390">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I should add, it would be great if we could add some logic to the "generateData" / related utils to immediately transform the object literals translated from JSON into instances of model classes / objects. That way we have some type consistency throughout the app, rather than various components handling the translation on their own. I think whether we use Immutable.js is a separate issue, really just an implementation detail.</p>
</pre>
</details>

</ul>
</details>

<p>I should add, it would be great if we could add some logic to the "generateData" / related utils to immediately transform the object literals translated from JSON into instances of model classes / objects. That way we have some type consistency throughout the app, rather than various components handling the translation on their own. I think whether we use Immutable.js is a separate issue, really just an implementation detail.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=264610&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-264610">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Update on what we did: <a href="https://github.com/tfennelly/blueocean-plugin/blob/JENKINS-36439/redux/README.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/tfennelly/blueocean-plugin/blob/JENKINS-36439/redux/README.md</a></p>

<p>Work on this has been parked for now.</p>
</pre>
</details>

</ul>
</details>

<p>Update on what we did: <a href="https://github.com/tfennelly/blueocean-plugin/blob/JENKINS-36439/redux/README.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/tfennelly/blueocean-plugin/blob/JENKINS-36439/redux/README.md</a></p>

<p>Work on this has been parked for now.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tfennelly.png" title="tfennelly's avatar" /> tfennelly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-36439?focusedId=265326&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-265326">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Closing this as Cliff is already working on introducing MobX in the personalization.</p>
</pre>
</details>

</ul>
</details>

<p>Closing this as Cliff is already working on introducing MobX in the personalization.</p>
